source/s3_client.c

```diff
@@ -336,6 +342,29 @@ PyObject *aws_py_s3_client_new(PyObject *self, PyObject *args) {
     s3_client->py_core = py_core;
     Py_INCREF(s3_client->py_core);
 
+    if (network_interface_names_py != Py_None) {
+        if (!PySequence_Check(network_interface_names_py)) {
+            PyErr_SetString(PyExc_TypeError, "Expected network_interface_names to be a sequence.");
+            goto cleanup;
+        }
+        Py_ssize_t listSize = PySequence_Size(network_interface_names_py);
+        num_network_interface_names = (size_t)listSize;
+        network_interface_names =
+            aws_mem_calloc(allocator, num_network_interface_names, sizeof(struct aws_byte_cursor));
+        for (Py_ssize_t i = 0; i < num_network_interface_names; ++i) {
+            PyObject *strObj = PySequence_GetItem(network_interface_names_py, i);
```

Comment:
Though python is generally excellent. You really need to read the docs for every single function in the c-api. Behavior is wildly inconsistent: [PyList_GetItem()](https://docs.python.org/3/c-api/list.html#c.PyList_GetItem) returned a borrowed reference, but `PySequence_GetItem()` returns a **new** reference.

I always put a `/* New reference */` or `/* Borrowed reference*/` comment after functions that return a `PyObject *`, to try and make this clear
```suggestion
            PyObject *strObj = PySequence_GetItem(network_interface_names_py, i); /* New reference */
```

or use the [PySequence_Fast](https://docs.python.org/3.11/c-api/sequence.html#c.PySequence_Fast) API, which will only do 1 new reference, instead of N (not that performance matters here, just do whatever makes the simplest cleanup code)

Reply:
Thanks, I have updated it with the comment and decRef. We can't use the fast API because https://docs.python.org/3.11/c-api/sequence.html#c.PySequence_Fast_GET_ITEM is not part of the stable API.

Reply:
Why is PySequence_Fast() in the Stable ABI, but none of the functions you can call on that object are ?!? 🤷‍♀️

keep it weird Python